### ISSUE TITLE
Empty Go structs lead to invalid JSON objects once marshalled

### ISSUE THREAD

**Issue by sebgl** [elastic/cloud-on-k8s#5544]
If I create a minimal `Kibana` object as code (in Go), the apiserver will reject my object because it's invalid:

```
{"message":"Kibana.kibana.k8s.elastic.co \"kb\" is invalid: spec.enterpriseSearchRef.name: Required value"}
```

The code does not create the `enterpriseSearch` field, it gets created automatically because the Kibana struct defines it. Even if the json tags specify `omitempty`, the JSON serialization of my objects ends up with `spec.enterpriseSearchRef: {}` (that's a struct, not a pointer, in the case of a pointer it wouldn't get rendered). That JSON is invalid, because `enterpriseSearchRef.name` is a required field.

When manipulating yaml by hand (rather than Go structs), things are a bit easier because we can choose not to write the `enterpriseSearchRef: {}` field. 

This makes Go developers life a bit trickier, as we now not only need to think about populating the fields we care about, but also about zeoing-out properly the fields we don't care about. In that particular case it's even trickier: if we set `enterpriseSearch.name` to `""`, that field will not get JSON-rendered because of the `omitempty` tag on the `name` field.

Potential solutions I see to this:
1. Make `enterpriseSearchRef` a pointer. So if not populated, it does not get JSON-rendered.
2. Do not make `enterpriseSearchRef.name` a required field. So if not specified the JSON object is still valid.
3. Do nothing and let the Go devs be responsible of doing the right thing: actually not possible in that case with the combination of the empty name that won't get rendered.

1) is my favourite option, because it leads to generating JSON files that only include the minimal fields required. It does mean we'd get an additional pointer in the code though, which isn't great.

There are probably more fields in our various CRDs that cause this problem. This is the first I experience.

**Comment by sebgl** [comment]
Hmm actually solution 2) has been implemented 3 days ago: https://github.com/elastic/cloud-on-k8s/pull/5240/files#diff-3f756fbb144c4b1ae321b23d15fb41c24bfbf76854425c492cb7b01f326eaa21R101

(solving my immediate problem - but I think there's still a discussion to have about option 1 and not rendering optional fields)

**Comment by pebrc** [comment]
We did some more digging and found that this particular problem was caused by a discrepancy in client and server version. The client was based on a more recent version of ECK that no longer required this value but the server version was the last released version. If server and client versions align this problem does not present itself. We are programmatically creating Kibana's through Go code all the time in our e2e tests.
